### PR TITLE
make the deployment control loop generic

### PR DESCRIPTION
**Pull request by deads2k**
This is needed before start promoting to library-go for re-use in controller manager

### CODE REVIEW

**Comment by deads2k** [comment]
now with more bug fixes.

**Comment by deads2k** [comment]
@sttts Once we get this in, I'll move the bits to library-go and pull them down into the kube-controller-manager.

**Comment by sttts** [comment]
This is really hard to review. Github shows removal and creation of tons of files. Any way to split code moves and code changes?

File: pkg/apis/kubeapiserver/v1alpha1/types.go [hunk 1]
```diff
@@ -1,6 +1,7 @@
 package v1alpha1
 
 import (
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator/staticpodcontroller"
```
Comment (soltysh) [inline pkg/apis/kubeapiserver/v1alpha1/types.go]: Down to openshift imports :stuck_out_tongue_closed_eyes:

File: pkg/operator/staticpod_interface.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	instance, err := c.informers.Kubeapiserver().V1alpha1().KubeAPIServerOperatorConfigs().Lister().Get("instance")
+	if err != nil {
+		return nil, nil, "", err
+	}
+
+	return &instance.Spec.OperatorSpec, &instance.Status.StaticPodConfigStatus, instance.ResourceVersion, nil
+}
+
+func (c *staticPodOperatorClient) UpdateStatus(resourceVersion string, status *staticpodcontroller.StaticPodConfigStatus) (*staticpodcontroller.StaticPodConfigStatus, error) {
+	original, err := c.informers.Kubeapiserver().V1alpha1().KubeAPIServerOperatorConfigs().Lister().Get("instance")
```
Comment (soltysh) [inline pkg/operator/staticpod_interface.go]: You're swallowing the error, why not using `CurrentStatus` here also?

File: pkg/operator/staticpodcontroller/installer_controller.go [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+		ret.TargetDeploymentID = 0
+
+		errors := []string{}
+		for _, containerStatus := range installerPod.Status.ContainerStatuses {
+			if containerStatus.State.Terminated != nil && len(containerStatus.State.Terminated.Message) > 0 {
+				errors = append(errors, fmt.Sprintf("%q: %s", containerStatus.Name, containerStatus.State.Terminated.Message))
+			}
+		}
+		if len(errors) == 0 {
+			errors = append(errors, "no detailed termination message, see `oc get -n %q pods/%q -oyaml`", installerPod.Namespace, installerPod.Name)
```
Comment (soltysh) [inline pkg/operator/staticpodcontroller/installer_controller.go]: Maybe also suggest seeing events?

File: pkg/operator/staticpodcontroller/installer_controller.go [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+}
+
+func getInstallerPodName(deploymentID int32, nodeName string) string {
+	return fmt.Sprintf("installer-%d-%s", deploymentID, nodeName)
+}
+
+// createInstallerPod creates the installer pod with the secrets required to
+func (c *InstallerController) createInstallerPod(currKubeletState *KubeletState, operatorSpec *operatorv1alpha1.OperatorSpec, deploymentID int32) (*KubeletState, error) {
+	required := resourceread.ReadPodV1OrDie(v311_00_assets.MustAsset("v3.11.0/kube-apiserver/installer-pod.yaml"))
+	switch corev1.PullPolicy(operatorSpec.ImagePullPolicy) {
```
Comment (soltysh) [inline pkg/operator/staticpodcontroller/installer_controller.go]: Why we care about this validation here? Or if we do validate why not other fields too?

File: pkg/operator/staticpodcontroller/interfaces.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+	"k8s.io/client-go/tools/cache"
+)
+
+type OperatorClient interface {
+	Informer() cache.SharedIndexInformer
+	Get() (*operatorv1alpha1.OperatorSpec, *StaticPodConfigStatus, string, error)
+	UpdateStatus(string, *StaticPodConfigStatus) (*StaticPodConfigStatus, error)
+}
+
+// TODO move to openshift/api I think
```
Comment (soltysh) [inline pkg/operator/staticpodcontroller/interfaces.go]: Why? IIUC this is only for kube-apiserver, is it not?

File: pkg/operator/staticpod_interface.go (hunk 2 above)
Comment (deads2k) [reply pkg/operator/staticpod_interface.go]: > You're swallowing the error, why not using `CurrentStatus` here also?

you need a full object of some kind to build the update payload until we switch to a dynamic client.

File: pkg/operator/staticpodcontroller/installer_controller.go (hunk 4 above)
Comment (deads2k) [reply pkg/operator/staticpodcontroller/installer_controller.go]: > Why we care about this validation here? Or if we do validate why not other fields too?

Shrugs. @mfojtik's seems reasonable.

File: pkg/operator/staticpodcontroller/interfaces.go (hunk 5 above)
Comment (deads2k) [reply pkg/operator/staticpodcontroller/interfaces.go]: > Why? IIUC this is only for kube-apiserver, is it not?

apiserver, controller, scheduler, possibly etcd.

**Comment by deads2k** [comment]
pull updated on latest, generic, library-go

**Comment by sttts** [comment]
Now this is much better to read.

/lgtm

Needs rebase.

**Pull request merged**